File: problem-types/equality_of_matches.py
import json
import sys

def entry_form(data, kwargs):
	match_length = 58  # длина спички
	match_width = 7  # ширина спички
	match_width_ = 10 # ширина картинки спички
	pad_ = (match_width_ - match_width) / 2  # сдвиг между картинкой и тенью
	match_length_ = 58 # длина картинки спички
	full_match = match_length + match_width  # длина с отступом для удобства
	full_match_ = match_length_ + match_width_  # длина картинки с отступом для удобства
	num_height = match_length * 2 + match_width * 3  # высота цифр
	num_width = match_length + match_width * 2  # ширина цифр
	symb_pad = 25  # отступ между символами в выражении
	margin = {'left': 20, 'top': 30, 'right': 20, 'bottom': 30}  # общие отступы
	num_amount = len(data['nums'])
	sgn_amount = len(data['sgn'])
	symb_amount = sgn_amount + num_amount
	rel_width, rel_height = (50, 50)
	view_box = {
		'height': num_height + margin['top'] + margin['bottom'],
		'width': num_width * num_amount + 
				sgn_amount * match_length + 
				symb_pad * (symb_amount - 1) + 
				margin['left'] + margin['right'] +
				rel_width
	}
	grid = [
		['hor', match_width, 0, match_length, match_width],
		['vert', 0, match_width, match_width, match_length],
		['vert', full_match, match_width, match_width, match_length],
		['hor', match_width, full_match, match_length, match_width],
		['vert', 0, full_match + match_width, match_width, match_length],
		['vert', full_match, full_match + match_width, match_width, match_length],
		['hor', match_width, full_match * 2, match_length, match_width]
	]  # общая разметка rect под цифры

	num_grid = [
		{'dir': 'hor', 'x':  match_width - pad_, 'y': match_width, 'width': match_width_, 'height': match_length_, 'transform': f'rotate(-90 {match_width} {match_width})'},
		{'dir': 'vert', 'x': -pad_, 'y': match_width, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(0)'},
		{'dir': 'vert', 'x': full_match - pad_, 'y': match_width, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(0)'},
		{'dir': 'hor', 'x':  match_width - pad_, 'y': full_match + match_width, 'width': match_width_, 'height': match_length_, 'transform': f'rotate(-90 {match_width} {match_width + full_match})'},
		{'dir': 'vert', 'x': -pad_, 'y': full_match + match_width, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(0)'},
		{'dir': 'vert', 'x': full_match - pad_, 'y': full_match + match_width, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(0)'},
		{'dir': 'hor', 'x':  match_width - pad_, 'y': full_match * 2 + match_width, 'width': match_width_, 'height': match_length_, 'transform': f'rotate(-90 {match_width} {full_match * 2 + match_width})'},
	]


	sgn_grid = {
		'-': [{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(90)'}],
		'+': [
				{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(90)'},
				{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(0)'}
			],
		'*': [
				{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(135)'},
				{'x': -match_width_ / 2, 'y': -match_length / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(45)'}
			],
		'=': [
				{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(90) translate(-15 0)'},
				{'x': -match_width_ / 2, 'y': -match_length_ / 2, 'width': match_width_, 'height': match_length_, 'transform': 'rotate(90) translate(15 0)'}
			]
	}  # разметка знаков

	translate = {
		0: [0, 1, 2, 4, 5, 6],
		1: [2, 5],
		2: [0, 2, 3, 4, 6],
		3: [0, 2, 3, 5, 6],
		4: [1, 2, 3, 5],
		5: [0, 1, 3, 5, 6],
		6: [0, 1, 3, 4, 5, 6],
		7: [0, 2, 5],
		8: [0, 1, 2, 3, 4, 5, 6],
		9: [0, 1, 2, 3, 5, 6]
	}  # индексы num_grid для отрисовки всех цифр

	step = data['step']
	if step == 0:
		data = kwargs['default']

	yield '<input name="answer" type="hidden" />'
	
	yield f'<svg version="1.1" width="{view_box["width"]}" height="{view_box["height"]}" overflow="visible" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">'
		
	yield '<defs>'
	yield '<filter id="sofGlow" height="110%" width="110%" x="-5%" y="-5%">'
	yield '<feMorphology operator="dilate" radius="4" in="SourceAlpha" result="thicken" />'
	yield '<feGaussianBlur in="thicken" stdDeviation="10" result="blurred" />'
	yield '<feFlood flood-color="rgb(0,186,255)" result="glowColor" />'
	yield '<feComposite in="glowColor" in2="blurred" operator="in" result="softGlow_colored" />'
	yield '<feMerge>'
	yield '<feMergeNode in="softGlow_colored"/>'
	yield '<feMergeNode in="SourceGraphic"/>'
	yield '</feMerge>'
	yield '</filter>'
	yield '</defs>'

	cur_dist = 0
	yield f'<g class="field" transform="translate({margin["left"]} {margin["top"]})">'
	for num_ind in range(num_amount):
		yield f'<g class="num" num="{num_ind}" transform="translate({cur_dist + symb_pad * (num_ind > 0)} {0})">'
		cur_dist += symb_pad * (num_ind > 0) + num_width
		for ind in range(len(grid)):
			yield f'<rect direction="{grid[ind][0]}" pos="{ind}" num="{num_ind}" x="{grid[ind][1]}" y="{grid[ind][2]}" width="{grid[ind][3]}" height="{grid[ind][4]}" />'
		
		if 'cur_config' in data.keys():
			print(data['cur_config'], file=sys.stderr)
			for ind in json.loads(data['cur_config'])[str(num_ind)]:
				rect = num_grid[int(ind)]
				yield f"""<image class="match active" num="{num_ind}" pos="{ind}" href="/static/problem_assets/match.svg" 
								direction="{rect["dir"]}"
								x="{rect["x"]}" y="{rect["y"]}" 
								width="{rect["width"]}" 
								height="{rect["height"]}" 
								preserveAspectRatio="none" 
								transform="{rect["transform"]}"/>"""
		else:
			for ind in translate[data['nums'][num_ind]]:
				rect = num_grid[ind]
				yield f"""<image class="match active" num="{num_ind}" pos="{ind}" href="/static/problem_assets/match.svg" 
								direction="{rect["dir"]}"
								x="{rect["x"]}" y="{rect["y"]}" 
								width="{rect["width"]}" 
								height="{rect["height"]}" 
								preserveAspectRatio="none" 
								transform="{rect["transform"]}"/>"""
		yield '</g>'
		if (num_ind < sgn_amount):
			yield f'<g class="sgn" type="{data["sgn"][num_ind]}" transform="translate({cur_dist + match_length / 2 + symb_pad} {num_height / 2})">'
			for rect in sgn_grid[data['sgn'][num_ind]]:
				yield f"""<image class="match passive" href="/static/problem_assets/match.svg" 
							x="{rect["x"]}" y="{rect["y"]}" 
							width="{rect["width"]}" 
							height="{rect["height"]}" 
							preserveAspectRatio="none" 
							transform="{rect["transform"]}"/>"""
			yield '</g>'
			cur_dist += symb_pad + match_length
	if step == 0:
		yield f'<rect class="hide_" style="display:none; fill:transparent" x="0" y="0" width="{view_box["width"] - rel_width}" height="{view_box["height"]}"/>'
	else:
		yield f'<rect class="hide_" style="display:block; fill:transparent" x="0" y="0" width="{view_box["width"] - rel_width}" height="{view_box["height"]}"/>'
	yield '</g>'
	yield f"""<image class="reload" href="/static/problem_assets/reload.png" 
			width={rel_width} height={rel_height}
			x={view_box["width"] - rel_width} y={view_box["height"] - rel_height - margin["bottom"]} />"""
	yield '</svg>'

def steps(step_num, params, data):
	if 'type' in params.keys():
		if params['type'] == 'move':
			if data['step'] > 1:
				return {'answer': 'not_accepted'}	
			else:
				data['step'] = 1
				data['cur_config'] = params['answer']
				return {'answer': 'accepted', 'data_update': data}
		elif params['type'] == 'reload':
			data['step'] = 0
			return {'answer': '', 'data_update': data}

def validate(data, ans):
	translate = {
		0: [0, 1, 2, 4, 5, 6],
		1: [2, 5],
		2: [0, 2, 3, 4, 6],
		3: [0, 2, 3, 5, 6],
		4: [1, 2, 3, 5],
		5: [0, 1, 3, 5, 6],
		6: [0, 1, 3, 4, 5, 6],
		7: [0, 2, 5],
		8: [0, 1, 2, 3, 4, 5, 6],
		9: [0, 1, 2, 3, 5, 6]
	}  # индексы num_grid для отрисовки всех цифр
	tmp = True
	ans = json.loads(ans)
	for num in range(len(data['correct'])):
		if set(translate[data['correct'][num]]) != set(map(int, ans[str(num)])):
			tmp = False
	return tmp

File: problem-types/equality_of_matches.js
function autoscroll(x, y) {
	block_nav()
	let [x_add, y_add] = [100, 100]
	let [y_tmp, x_tmp] = ['no', 'no']
	let dur = {'left': -1, 'right': 1, 'up': -1, 'down': 1, 'no': 0}
	$(scroll_p).css({'left': x, 'top': y - -y_add})
	if (!in_access_zone(scroll_p))
		y_tmp = 'down'

	$(scroll_p).css({'left': x, 'top': y - y_add})
	if (!in_access_zone(scroll_p))
		y_tmp = 'up'

	$(scroll_p).css({'top': y, 'left': x - -x_add})
	if (!in_access_zone(scroll_p))
		x_tmp = 'right'

	$(scroll_p).css({'top': y, 'left': x - x_add})
	if (!in_access_zone(scroll_p))
		x_tmp = 'left'

	scrollBy(dur[x_tmp] * x_add / 2, dur[y_tmp] * y_add / 2)
}

function in_window(obj) {
	let rect = $(obj)[0].getBoundingClientRect()
	let [hor_add, vert_add] = [rect.width / 2, rect.height / 2]
	if (rect.left + hor_add < 0 || rect.right - hor_add > $(window).width())
		return false
	if (rect.top + vert_add < 0 || rect.bottom - vert_add > $(window).height())
		return false
	return true
}

function is_intersect(obj1, obj2) {
	let [rect1, rect2] = [$(obj1)[0].getBoundingClientRect(), $(obj2)[0].getBoundingClientRect()]
	if (rect1.left > rect2.right || rect2.left > rect1.right) 
		return false
	if (rect1.bottom < rect2.top || rect2.bottom < rect1.top)
		return false
	return true
}

function in_access_zone(obj) {
	let nav = $('nav.user_nav')
	if (is_intersect(obj, nav))
		return false
	if (!in_window(obj))
		return false
	return true
}

function getSVGCoordinates(event) {
	if (event.touches) {
		let coordinatePoint = new DOMPoint(event.touches[0].clientX, event.touches[0].clientY)
		let svg = document.querySelector('svg')
		return coordinatePoint.matrixTransform(svg.getScreenCTM().inverse())
	}
	else {
		let coordinatePoint = new DOMPoint(event.clientX, event.clientY)
		let svg = document.querySelector('svg')
		return coordinatePoint.matrixTransform(svg.getScreenCTM().inverse())
	}
}

function occupied(rect) {
	let [pos, num] = [$(rect).attr('pos'), $(rect).attr('num')]
	let match = $(`.match.active[pos=${pos}][num=${num}]`)
	return (match.length > 0 && !match.hasClass('targeted'))
}

function dist(obj1, obj2) {
	let rect1 = obj1.getBoundingClientRect()
	let rect2 = obj2.getBoundingClientRect()
	let [x1, y1] = [(rect1.right + rect1.left) / 2, (rect1.bottom + rect1.top) / 2]
	let [x2, y2] = [(rect2.right + rect2.left) / 2, (rect2.bottom + rect2.top) / 2]
	return (x2 - x1) ** 2 + (y1 - y2) ** 2
}

function update_best() {
	$('rect').removeClass('best')
	let obj = $('.targeted')[0]
	let [best_ind, min_dist] = [0, 10 ** 18]
	$('rect').each(function(ind){
		let new_dist = dist(this, obj)
		if (new_dist < min_dist) {
			[best_ind, min_dist] = [ind, new_dist]
		}
	})
	if (min_dist <= min_dist_ && !occupied($(`rect:eq(${best_ind})`))) {
		$(`rect:eq(${best_ind})`).addClass('best')
	}
}

function start_move(e) {
	let obj = $(e.currentTarget)
	$('svg').append(obj)
	let svg = $('svg')[0].getBoundingClientRect()
	$(obj).attr({
			'x': getSVGCoordinates(e).x - match_width / 2, 
			'y': getSVGCoordinates(e).y - match_length / 2,
			'transform': 'rotate(0)'})
	$(obj).addClass('targeted')
	update_best()
	$(document).on('mousemove touchmove', move)
}

function move(e) {
	let obj = $('.targeted')
	let svg = $('svg')[0].getBoundingClientRect()
	$(obj).attr({
			'x': getSVGCoordinates(e).x - match_width / 2,
			'y': getSVGCoordinates(e).y - match_length / 2})
	let [x, y] = [e.clientX, e.clientY]
	if (!in_access_zone(obj))
		drop()
	else 
		update_best()
	autoscroll(e.clientX, e.clientY)
}

function back_to_drag(obj=$('.match.targeted:eq(0)')) {
	console.log($(obj).attr('num'))
	let [pos, num] = [$(obj).attr('pos'), $(obj).attr('num')]
	let rect = $(`rect[pos=${pos}][num=${num}]`)
	$(obj).attr('x', $(rect).attr('x') - pad_)
	$(obj).attr('y', $(rect).attr('y'))
	if ($(rect).attr('direction') === 'hor') {
		$(obj).attr('y', $(obj).attr('y') - -rect_width)
		$(obj).attr('transform', `rotate(-90 ${$(rect).attr('x')} ${$(obj).attr('y')})`)
	}
	$(obj).appendTo(`g.num[num=${num}]`)
}

function drop() {
	let obj = $('.targeted')
	if ($('.best').length) {
		let response = true
		let rect = $('.best')
		let [prev_num, prev_pos] = [$(obj).attr('num'), $(obj).attr('pos')]
		let [num, pos] = [$(rect).attr('num'), $(rect).attr('pos')]
		let same = (num == $(obj).attr('num') && pos == $(obj).attr('pos'))
		$(obj).attr({
			'x': $(rect).attr('x') - pad_,
			'y': $(rect).attr('y') - -rect_width * ($(rect).attr('direction') === 'hor'),
			'pos': pos,
			'num': num,
			})
		if ($(rect).attr('direction') === 'hor')
			$(obj).attr('transform', `rotate(-90 ${$(rect).attr('x')} ${$(obj).attr('y')})`)
		rect.removeClass('best')
		$(obj).appendTo(`g.num[num=${num}]`)
		$(obj).removeClass('targeted')

		if (!same) {
			let url = new URL(window.location.href + 'xhr')
			let solution = $('#problem_form')[0].outerHTML
			$.post(url, JSON.stringify({'type': 'move', 'answer': get_answer()}), function(data){
				if (data != 'accepted') {
					show_xhr('Больше перекладывать нельзя!')
					$(obj).attr({
						'pos': prev_pos,
						'num': prev_num,
						})
					back_to_drag($(obj))
				}
				else
					$('.hide_').css('display', 'block')
			})
		}
	}
	else
		back_to_drag()
	$(obj).removeClass('targeted')
	$(document).off('mousemove touchmove')
}

function get_answer() {
	let ans = {}
	$('g.num').each(function(ind){ans[$(this).attr('num')] = []})
	$('.match.active').each(function(ind){
		ans[$(this).attr('num')].push($(this).attr('pos'))
	})
	return JSON.stringify(ans)
}

const match_width = Math.min($('.match').attr('width'), $('.match').attr('height'))
const match_length = Math.max($('.match').attr('width'), $('.match').attr('height'))
const rect_width = $('rect').attr('height')
const pad_ = (match_width - rect_width) / 2
const min_dist_ = 900 //квадрат расстояния между rect и match для подстветки

var scroll_p = $(document.createElement('div')).addClass('scroll_div')
$('body').append(scroll_p)

$('.match.active').on('mousedown touchstart', start_move)
$('.match.active').on('mouseup touchend', drop)

let url = new URL(window.location.href + 'xhr')
$('.reload').on('click touchstart', function(){
	$.post(url, JSON.stringify({'type': 'reload'}), function(data){window.location.reload("true")})})

$('button.submit_button').on('click touchstart', 
	function(){$('input[name="answer"]').val(get_answer())})